Libraries/LibC/time.cpp

```diff
@@ -106,53 +107,222 @@ time_t mktime(struct tm* tm)
     for (int month = 0; month < tm->tm_mon; ++month)
         tm->tm_yday += __days_per_month[month];
+    if (tm->tm_mon > 1 && __is_leap_year(1900 + tm->tm_year))
+        ++tm->tm_yday;
+
     days += tm->tm_yday;
-    return days * __seconds_per_day + seconds;
+    return days * __seconds_per_day + seconds + timezone;
 }
 
 struct tm* localtime(const time_t* t)
 {
-    if (!t)
-        return nullptr;
     static struct tm tm_buf;
     return localtime_r(t, &tm_buf);
 }
 
+struct tm* localtime_r(const time_t* t, struct tm* tm)
+{
+    if (!t)
+        return nullptr;
+    time_to_tm(tm, (*t) - timezone);
+    return tm;
+}
+
 struct tm* gmtime(const time_t* t)
 {
-    // FIXME: This is obviously not correct. What about timezones bro?
-    return localtime(t);
+    static struct tm tm_buf;
+    return gmtime_r(t, &tm_buf);
 }
 
 struct tm* gmtime_r(const time_t* t, struct tm* tm)
 {
-    // FIXME: This is obviously not correct. What about timezones bro?
-    return localtime_r(t, tm);
+    if (!t)
+        return nullptr;
+    time_to_tm(tm, *t);
+    return tm;
 }
 
-char* asctime(const struct tm*)
+static char wday_short_names[7][4] = {
+    "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
+};
+
+static char wday_long_names[7][10] = {
+    "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"
+};
+
+static char mon_short_names[12][4] = {
+    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
+    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
+};
+
+static char mon_long_names[12][10] = {
+    "January", "February", "March", "April", "May", "June",
+    "July", "Auguest", "September", "October", "November", "December"
+};
+
+char* asctime(const struct tm* tm)
 {
-    ASSERT_NOT_REACHED();
+    static char result[69];
+    snprintf(result, sizeof result, "%.3s %.3s %2d %02d:%02d:%02d %4d\n",
+        wday_short_names[tm->tm_wday],
+        mon_short_names[tm->tm_mon],
+        tm->tm_mday, tm->tm_hour,
+        tm->tm_min, tm->tm_sec,
+        1900 + tm->tm_year);
+    return result;
 }
 
-size_t strftime(char* destination, size_t, const char*, const struct tm*)
+//FIXME: Some formats are not supported.
+size_t strftime(char* destination, size_t max_size, const char* format, const struct tm* tm)
 {
-    // FIXME: Stubbed function to make nasm work. Should be properly implemented
-    strcpy(destination, "strftime_unimplemented");
-    return strlen("strftime_unimplemented");
+    int destination_pos = 0;
+    const int format_len = strlen(format);
+
+    for (int i = 0; i < format_len; ++i) {
+        if (format[i] != '%') {
+            if (destination_pos + 1 > static_cast<int>(max_size) - 1)
+                return 0;
+            destination[destination_pos++] = format[i];
+        } else if (++i < format_len) {
+            char component[30];
+
+            switch (format[i]) {
+            case 'a':
+                snprintf(component, sizeof component, "%s", wday_short_names[tm->tm_wday]);
+                break;
```

Comment:
Let's avoid C-style programming and use `StringBuilder` for this task instead. Something like this:

```cpp
StringBuilder builder;
...
case 'a':
    builder.append(wday_short_names[tm->tm_wday]);
    break;
...
case 'T':
    builder.appendf("%02d:%02d:%02d", ...);
    break;
...
```